### PR TITLE
Improving Test Coverage

### PR DESCRIPTION
**Pull request by thepacketgeek**
Adding unittest coverage and improving test experience:
- Adding more tests for flowspec, NLRI, and attributes
- Slow tests (ripe, mrt reads) are ignored for `$ cargo test` to reduce test time
- `--ignored` is added to the Github Actions CI to make sure these longer tests are included

### CODE REVIEW

**Comment by DevQps** [comment]
Looks good! I checked the whole code and I think it looks fine! The only things I found was the dbg! statement and that I am wondering what the best name for the Notification::with_data method should be. I wonder if there is some kind of convention for naming multiple constructors in Rust. Any idea?

**Comment by DevQps** [comment]
Nevermind... I read this:

```
You cannot have two methods with the same name. The common pattern is to have multiple methods (e.g. new, with_x, from_y), or use the Builder pattern if the number of combinations gets out of hand.
```

I am learning everyday I guess! Personally I would be more fan of `from_data` but either is fine for me. Thanks a lot again for your hard work!

**Comment by thepacketgeek** [comment]
Thank you! I still have some work to do for open.rs and attributes.rs. Will checkout the dog statement and use ‘from_data’

File: .github/workflows/CI.yml [hunk 1]
```diff
@@ -84,7 +85,7 @@ jobs:
       - name: Run cargo-tarpaulin
         uses: actions-rs/tarpaulin@v0.1
         with:
-          args: "--verbose --all-features --workspace --timeout 1200"
+          args: "--verbose --all-features --run-types Doctests,Tests --timeout 1200 --ignored"
```
Comment (thepacketgeek) [inline .github/workflows/CI.yml]: cargo-tarpaulin doesn't include doctests or ignored tests by default, so adding explicit `--run-types` and `--ignored`

File: src/lib.rs [hunk 2]
```diff
@@ -5,7 +5,8 @@
 //! # Examples
 //!
 //! ## Reading a MRT file containing BPG4MP messages
-//! ```
+//!
+//! ```no_run
```
Comment (thepacketgeek) [inline src/lib.rs]: Both of the lib.rs example doctests are run as part of the main test suite in `tests/ripe.rs`, so excluding from run (as it adds a significant amount of time to do each test twice)

File: src/open.rs [hunk 3]
```diff
@@ -256,14 +255,16 @@ impl OpenCapability {
                 cap_buf.write_u8(0)?; // Capability Length
             }
             OpenCapability::OutboundRouteFiltering(orfs) => {
-                let length = orfs.len();
+                cap_buf.write_u8(3)?; // Capability Type
```
Comment (thepacketgeek) [inline src/open.rs:258]: The type & length was missing, unittests caught this :)

**Comment by DevQps** [comment]
Long live unit testing :) Good work!

The only thing I was wondering is: If the doc test is marked as `--no_run`, will it still be executed when `--ignored` is executed? I understand the rationale for not wanting to run the same test twice. That is a bit ugly I guess. On the other hand: I once had a project where I had a similar construction. When the codebase changed I updated the file in tests but forgot the doctest which caused the doctest to be outdated without anyone knowing (it wasn't executed because `--no_run` back then).

What do you think? Is this an acceptable risk? Or should we run the tests? Or maybe remove one of the two tests? Or mork them differently somehow.

I am curious about your opinion on this.

**Comment by DevQps** [comment]
We could potentially also mark them as ignored, but I guess that's about all the approaches that we have.

**Comment by thepacketgeek** [comment]
When a doctest is marked `no_run`, it will be compiled but not run under any case (normal test runs, or with --ignored). If the tests were marked `ignored`, they would run with the --ignored flag, and we'd be back in the double-run situation.

We could mark the lib.rs doctests as ignored (since they're slower, only run on-demand and in CI), and remove the `ripe.rs` duplicate tests?  That has the advantage of making sure the examples continue to run, which I think is important to preserve

**Comment by thepacketgeek** [comment]
@DevQps I consolidated the duplicate tests, leaving the lib.rs doctest versions (ignored by default) to confirm that they will continue to work properly 👍 

I also fixed some tests in open.rs I hadn't finished, for a final cover bump over the 80% threshold 😃 (well, locally anyway, i can't figure out why codecov is dropping the coverage for notification.rs):

```
|| Tested/Total Lines:
|| src/lib.rs: 117/137 +0%
|| src/notification.rs: 33/39 +0%
|| src/open.rs: 216/280 +21.44110275689224%
|| src/update/attributes.rs: 320/462 +0%
|| src/update/flowspec.rs: 215/255 +0%
|| src/update/mod.rs: 152/201 +0.12189054726368109%
|| src/update/nlri.rs: 121/143 +0%
|| src/util.rs: 56/62 +0%
|| tests/common/mod.rs: 64/79 +0%
|| tests/decode.rs: 95/100 +0%
|| tests/encode.rs: 164/164 +0%
|| tests/flowspec.rs: 129/155 +0%
|| tests/old_bugs.rs: 6/6 +0%
|| tests/pcap.rs: 20/24 +0%
||
81.06% coverage, 1708/2107 lines covered, +2.2899973150193915% change in coverage
```

**Comment by DevQps** [comment]
Super cool! I totally agree on the testing part btw! Good solution I think!

Nice that the coverage has been increased so much as well! It's really amazing that we're slowly crawling up to that 100%! I think 80% is already sonething we (mostly you :)) can be proud of!

**Comment by DevQps** [comment]
I am wondering what it could be indeed... If are not quicker then me I might have some time to look at it on Saturday!

**Comment by thepacketgeek** [comment]
@DevQps whew, finally got it! I could tell from the percentage difference and Actions logs that the doctests were not being run. Found out that the actions/cargo-tarpaulin was getting deprecated and I believe it doesn't support doctest, so went down a rabbit hole and finally got the coverage tests working

**Comment by DevQps** [comment]
@thepacketgeek Nice that you were able to solve it! Glad that the nightly version works however! Would you like me to review it one more time and merge it if I have no more comments? Thanks alot for the great work!

**Comment by thepacketgeek** [comment]
@DevQps , yes, please look over again as additional eyes are helpful :) I can squash and merge with a more helpful summary of changes after you're happy with it.

**Comment by DevQps** [comment]
Aside these last comments I have nothing to add! I saw that in 2 tests there were some eprintln! statements, but I don't think this is any problem right? It might even help debugging if it would fail. 

If you think these points I raised are not valid, feel free to squash and merge!

**Comment by DevQps** [comment]
Ps. Sorry for the later response sometimes. Sometimes weeks can get pretty busy here. But I never forget it (or at least I hope I never do). Feel free to spam me anytime you think I did forget it. Soon I will be on vacation (24 July), but before that I hope to spend some Wednesdays working on it myself as well.

**Comment by thepacketgeek** [comment]
> I saw that in 2 tests there were some eprintln! statements, but I don't think this is any problem right? It might even help debugging if it would fail.

Exactly! Those will only display if the test fails, and it will only print the items that failed the roundtrip (so it's clear what the difference is)

**Pull request merged**